Preliminary GPU port of MOOSE using Kokkos

NamjaeChoi opened the pull request:
Refs #30655.

- [x] the API names in the `KokkosMesh`
- [x] Adding errors for adaptivity/displaced mesh
- [x] non-hacky copy construction
- [x] Syntax documentation

NamjaeChoi committed:
Kokkos build scripts #30655

NamjaeChoi committed:
GPU base header and initialization #30655

NamjaeChoi committed:
GPU basic data classes #30655

NamjaeChoi committed:
GPU mesh data #30655

NamjaeChoi committed:
Add initialize parameter to MOOSE classes to avoid their initialization during copy #30655

NamjaeChoi committed:
GPU assembly class #30655

NamjaeChoi committed:
GPU system classes #30655

NamjaeChoi committed:
Initialize GPU assembly and systems #30655

NamjaeChoi committed:
GPU datum object to privatize thread-wise data #30655

NamjaeChoi committed:
GPU variable coupleable system #30655

NamjaeChoi committed:
GPU tests #30655

NamjaeChoi committed:
GPU block and boundary restrictables #30655

NamjaeChoi committed:
GPU kernels and boundary conditions #30655

NamjaeChoi committed:
Framework-level GPU kernels and boundary conditions #30655

NamjaeChoi committed:
Module-level GPU kernels and boundary conditions #30655

NamjaeChoi committed:
Add GPU kernels and boundary conditions to the system #30655

NamjaeChoi committed:
GPU material system #30655

NamjaeChoi committed:
Framework-level GPU materials #30655

NamjaeChoi committed:
Module-level GPU materials #30655

NamjaeChoi committed:
Add GPU materials to the system #30655

NamjaeChoi committed:
Establish GPU calculation flow #30655

loganharbour commented on framework/include/base/Assembly.h:
```diff
@@ -223,6 +223,15 @@ class Assembly
+   * Returns the reference to the quadrature of specified block and dimension
+   * @return A _reference_ to the unique pointer.  Make sure to store this as a reference!
+   */
+  const std::unique_ptr<libMesh::QBase> & writeableQRule(unsigned int dim, SubdomainID block)
```
Why return a reference to a unique_ptr? That's an implementation detail that is not typically exposed. Same with `writeableQRuleFace`.

NamjaeChoi replied:
Changed to return a pointer.

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -1565,6 +1565,21 @@ class MooseApp : public ConsoleStreamInterface,
+  bool hasGPUs() { return _has_GPUs > 0; }
+
+private:
+  bool _has_GPUs = false;
```
Initialize in the initializer list unless you have an explicit reason not to do so. `has_GPUs` -> `has_gpus`.

However, we need to have a better discussion on device discovery and how we can unify it across libraries (kokkos, libtorch, neml2). It would be easy to be confused about "has gpus" working for libtorch and neml2.

Regardless, something this simple really needs a docstring due to the ambiguity. Same for the getter.

NamjaeChoi replied:
That's a good point. It can actually conflict with other libraries. We need a unified method.

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -1565,6 +1565,21 @@ class MooseApp : public ConsoleStreamInterface,
+private:
+  bool _has_GPUs = false;
+
+#ifdef MOOSE_HAVE_GPU
```
```suggestion
I believe this should be `MOOSE_HAVE_KOKKOS` for specificity. 
```

NamjaeChoi replied:
I can change that.

NamjaeChoi replied:
Changed.

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -1565,6 +1565,21 @@ class MooseApp : public ConsoleStreamInterface,
+   */
+public:
+  /// Return whether GPUs exist
+  bool hasGPUs() { return _has_GPUs > 0; }
```
```suggestion
  bool hasGPUs() { return _has_GPUs; }
```

loganharbour commented on framework/include/base/MooseObject.h:
```diff
@@ -30,7 +30,7 @@ class MooseObject : public ParallelParamObject, public std::enable_shared_from_t
   static InputParameters validParams();
 
-  MooseObject(const InputParameters & parameters);
+  MooseObject(const InputParameters & parameters, bool initialize = true);
```
I'll need to dig further into why this is needed. But changing the signature for the base object is something that concerns me. We're also really trying to not encourage adding more defaults.

This could be handled through private parameters.

loganharbour replied:
A `MooseObject` is used for many other things than just the residual objects that you have written. Making this capability available to _all other systems_ is dangerous and unwise

NamjaeChoi replied:
Check out the new algorithm.

loganharbour commented on framework/include/gpu/actions/AddGPUBCAction.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "MooseObjectAction.h"
+
+class AddGPUBCAction : public MooseObjectAction
```
Docstring

loganharbour commented on framework/include/gpu/actions/AddGPUKernelAction.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "MooseObjectAction.h"
+
+class AddGPUKernelAction : public MooseObjectAction
```
Docstring

loganharbour commented on framework/include/gpu/actions/AddGPUMaterialAction.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "MooseObjectAction.h"
+
+class AddGPUMaterialAction : public MooseObjectAction
```
Docstring

loganharbour commented on framework/include/gpu/actions/AddGPUNodalKernelAction.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "MooseObjectAction.h"
+
+class AddGPUNodalKernelAction : public MooseObjectAction
```
Docstring

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+
+#pragma once
+
+#ifdef MOOSE_GPU_SCOPE
```
This precompiler variable should probably also be named something better; I'm not really sure what it means upon first glance

NamjaeChoi replied:
Open to suggestions. This preprocessor is defined for the header/source files that are compiled by the GPU compiler.

NamjaeChoi replied:
Changed to `MOOSE_KOKKOS_SCOPE`.

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+
+void KokkosFree(void * ptr);
+
+enum class GPUMemcpyKind
```
Everything below needs to be put in a namespace and needs docstrings.

NamjaeChoi replied:
Put in `Moose::Kokkos` namespace. Docstring on its way.

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+  void createHost(const std::vector<uint64_t> n) { createInternal<true, false>(n); }
+  // Create array on device
+  void createDevice(const std::vector<uint64_t> n) { createInternal<false, true>(n); }
+  // Simply alias host data pointer for an initialized array
+  void aliasHost(T * ptr);
+  // Simply alias device data pointer for an initialized array
+  void aliasDevice(T * ptr);
+  // Simply allocate host data storage for an initialized array
+  void allocHost();
+  // Simply allocate device data storage for an initialized array
+  void allocDevice();
```
What does "simply" mean here?

NamjaeChoi replied:
Does the new docstring make sense?

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+#endif
+};
+
+template <typename T, unsigned int dimension>
```
Do the definitions need to go into a header file? Do we not have known types for what we will construct this with?

NamjaeChoi replied:
I support `GPUArray` to be of arbitrary type.

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+  bool _is_host_alias = false;
+  // Flag whether device data pointer is an alias of another pointer
+  bool _is_device_alias = false;
+  // Host data pointer
+  T * _host_data = nullptr;
+  // Device data pointer
+  T * _device_data = nullptr;
```
Not possible to use managed pointers for any of this data?

NamjaeChoi replied:
Managed pointers are not accessible on GPU.

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+  iterator end() const { return iterator(_host_data + _size); }
+
+private:
+  // Reference counter
```
Can these not be constructed in an initializer list?

loganharbour commented on framework/include/gpu/base/GPUAssembly.h:
```diff
@@ -0,0 +1,282 @@
+class FEProblemBase;
+class Assembly;
+
+class GPUAssembly : public GPUMeshHolder
```
Docstrings.

Long function definitions should be put at the end of the file.

NamjaeChoi replied:
Put in `Moose::Kokkos` namespace.

NamjaeChoi replied:
Put in `Moose::Kokkos` namespace.

loganharbour commented on framework/include/gpu/base/GPUAssembly.h:
```diff
@@ -0,0 +1,282 @@
+
+#ifdef MOOSE_GPU_SCOPE
+  // Get the FE type number
+  auto getFETypeNum(FEType type) const { return _fe_type_map.at(type); }
```
Needs error checking or at least libmesh_map_find

loganharbour reviewed (changes requested):
I've only gotten through a small portion - but there is enough to start with here.

I have one large question - do all of the `KOKKOS_FUNCTION` return methods need to go into a header? I'm very concerned with how much content is in headers instead of sources.

I'm also concerned with the length of this PR. Part of me would prefer that we have no changes to modules yet because the time required for review is so significant that I would prefer to have as small of a PR as possible.

hugary1995 commented:
I have one high-level comment.

If we can make the following work with the kokkos data structure:
```cpp
Kokkos::View<float*, Kokkos::CudaSpace> kokkos_view("data", N);

auto data_ptr = kokkos_view.data();  // This is a raw CUDA device pointer

torch::Tensor tensor = torch::from_blob(
    data_ptr,
    {N},
    torch::TensorOptions().dtype(torch::kFloat64).device(torch::kCUDA)
);
```
then porting kokkos moose to neml2 will be a near zero-shot effort.

NamjaeChoi commented:
> I have one large question - do all of the `KOKKOS_FUNCTION` return methods need to go into a header? I'm very concerned with how much content is in headers instead of sources.

The GPU methods should be inlinable unless it is compiled as relocatable device code (RDC). The RDC option is turned off by default, and I suggested the PETSc team to support turning it on, but they said it will take a non-trivial effort for the PETSc team to support turning it on. Also, inlining has non-negligible performance advantage on GPU.

NamjaeChoi commented:
> I have one high-level comment.
> 
> If we can make the following work with the kokkos data structure:
> 
> ```c++
> Kokkos::View<float*, Kokkos::CudaSpace> kokkos_view("data", N);
> 
> auto data_ptr = kokkos_view.data();  // This is a raw CUDA device pointer
> 
> torch::Tensor tensor = torch::from_blob(
>     data_ptr,
>     {N},
>     torch::TensorOptions().dtype(torch::kFloat64).device(torch::kCUDA)
> );
> ```
> 
> then porting kokkos moose to neml2 will be a near zero-shot effort.

I'm not using Kokkos views due to its strictness and I'm playing with low-level Kokkos APIs (i.e. I'm dealing with raw pointers).

NamjaeChoi commented:
> Part of me would prefer that we have no changes to modules yet because the time required for review is so significant that I would prefer to have as small of a PR as possible.

The changes in modules is minimal and can be taken out from this PR.

NamjaeChoi commented:
> > I have one high-level comment.
> > If we can make the following work with the kokkos data structure:
> > ```c++
> > Kokkos::View<float*, Kokkos::CudaSpace> kokkos_view("data", N);
> > 
> > auto data_ptr = kokkos_view.data();  // This is a raw CUDA device pointer
> > 
> > torch::Tensor tensor = torch::from_blob(
> >     data_ptr,
> >     {N},
> >     torch::TensorOptions().dtype(torch::kFloat64).device(torch::kCUDA)
> > );
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then porting kokkos moose to neml2 will be a near zero-shot effort.
> 
> I'm not using Kokkos views due to its strictness and I'm playing with low-level Kokkos APIs (i.e. I'm dealing with raw pointers). I'm wrapping the raw pointers with `GPUArray`.

hugary1995 commented:
> I'm not using Kokkos views due to its strictness and I'm playing with low-level Kokkos APIs (i.e. I'm dealing with raw pointers).

Are you at least using `kokkos_malloc`? It would still be trivial if you did something like
```cpp
double * device_ptr = static_cast<double*>(Kokkos::kokkos_malloc<Kokkos::CudaSpace>("my_data", N * sizeof(double)));
```

The three requirements are:
1. we have access to the raw cuda device pointer
2. we know the layout, i.e. size and stride
3. the allocated data outlives the torch tensor

After skimming through the PR, it seems to me that 2 and 3 are satisfied. I'm not sure about 1.

NamjaeChoi commented:
Yes, I'm using `kokkos_malloc`.

hugary1995 commented:
cool

loganharbour commented on framework/include/gpu/base/GPUDatum.h:
```diff
@@ -0,0 +1,180 @@
+#include "GPUSystem.h"
+#include "GPUVariable.h"
+
+class Datum
```
Docstrings. Move function definitions outside of class declaration. Needs to be namespaced

loganharbour commented on framework/include/gpu/base/GPUMap.h:
```diff
@@ -0,0 +1,84 @@
+#include <memory>
+#include <map>
+
+template <typename T1, typename T2>
```
Docstrings (at least for the class). Move function definitions outside of class declaration. Needs to be namespaced

loganharbour commented on framework/include/gpu/base/GPUMatrix.h:
```diff
@@ -0,0 +1,106 @@
+
+class GPUSystem;
+
+class GPUMatrix
```
Class docstring. Move function definitions outside of class declaration. Needs to e namespaced

loganharbour commented on framework/include/gpu/base/GPUReferenceWrapper.h:
```diff
@@ -0,0 +1,130 @@
+
+#include "MooseTypes.h"
+
+template <typename T>
```
Class docstring

loganharbour commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,987 @@
+  DEVICE_TO_DEVICE
+};
+
+template <typename T, unsigned int dimension = 1>
```
Needs to be namespaced

NamjaeChoi replied:
Changed.

loganharbour commented on framework/include/gpu/base/GPUThread.h:
```diff
@@ -0,0 +1,48 @@
+
+#include "MooseTypes.h"
+
+class GPUThread
```
Docstrings, definitions outside of declarations

loganharbour commented on framework/include/gpu/base/GPUVariable.h:
```diff
@@ -0,0 +1,55 @@
+
+class Coupleable;
+
+class GPUVariable
```
Class docstring

loganharbour commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+
+class GPUSystem;
+
+class GPUVector
```
Class docstring, namespace

loganharbour commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+  void close();
+
+public:
+  // GPU function tags
+  struct PackBuffer
+  {
+  };
+  struct UnpackBuffer
+  {
+  };
```
```suggestion
  /**
   * GPU function tags
   */
  ///@{
  struct PackBuffer
  {
  };
  struct UnpackBuffer
  {
  };
  ///@}
```

NamjaeChoi replied:
Changed.

loganharbour commented on framework/include/gpu/base/GPUResidualObject.h:
```diff
@@ -0,0 +1,217 @@
+                    bool nodal = false);
+  GPUResidualObject(const GPUResidualObject & object);
+
+  // GPU function tags
+  struct ResidualLoop
+  {
+  };
+  struct JacobianLoop
+  {
+  };
+  struct OffDiagJacobianLoop
+  {
+  };
```
```suggestion
  /**
   * GPU function tags
   */
  ///@{
  struct ResidualLoop
  {
  };
  struct JacobianLoop
  {
  };
  struct OffDiagJacobianLoop
  {
  };
  ///@}
```

NamjaeChoi replied:
Changed.

loganharbour commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+  GPUArray<PetscScalar> _local;
+  // Data that should be sent to ghost processes
+  GPUArray<PetscScalar> _ghost;
+  // Flag whether the vector will be assembled
+  bool _assemble = false;
+  // Flag whether the vector is ghosted
+  bool _is_ghosted = false;
+  // Flag whether the PETSc vector is a host vector
+  bool _is_host = false;
+  // Flag whether the vector was allocated
+  bool _is_alloc = false;
```
Defaults here also error prone. For example, you have a `bool assemble = false` default in create, but also a default here. It makes it difficult to see where these are actually set. You could construct this object, look at that construction, and not know what's actually going on

loganharbour commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+
+class GPUVector
+{
+private:
```
Why is public not defined first?

NamjaeChoi replied:
Is that you guys' guideline?

loganharbour commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+  void create(libMesh::NumericVector<PetscScalar> & vector,
+              const GPUSystem & system,
+              bool assemble = false);
+  /**
+   * Copy from/to the libMesh PetscVector
+   */
+  void copy(GPUMemcpyKind dir = GPUMemcpyKind::HOST_TO_DEVICE);
+  /**
+   * Free all the vector data
+   */
+  void destroy();
+  /**
+   * Restore the PETSc vector
+   */
+  void restore();
+  /**
+   * Assemble the vector
+   */
+  void close();
```
Where are these defined?

NamjaeChoi replied:
`GPUVector.K`

loganharbour reviewed (changes requested):
Through most of the base stuff.

I'd like @lindsayad to chime in on naming - folders, objects, helpers, etc. I think that everything should be prefixed with Kokkos. We have this for MFEM. Seeing that we have multiple ways to interact with GPUs, I don't think it's a good idea to call one specific implementation _the_ GPU implementation.

NamjaeChoi commented on framework/include/gpu/base/GPUMap.h:
```diff
@@ -0,0 +1,84 @@
+class GPUMap
+{
+private:
+  // Host map
+  std::shared_ptr<std::map<T1, T2>> _map;
```
I made GPU data classes (`GPUArray`, `GPUVector`, `GPUMap`, etc.) to be shallow-copied by default to avoid overhead during functor dispatch. If it's not `shared_ptr`, it will be deep-copied. I don't think `unique_ptr` is copy-constructible.

lindsayad commented on framework/include/gpu/base/GPUVector.h:
```diff
@@ -0,0 +1,157 @@
+
+class GPUSystem;
+
+class GPUVector
```
This must rely on defaulted constructors? I might make them explicitly defaulted in the class definition to help a developer/user grok this class

NamjaeChoi replied:
Actually, high-level users/developers will almost never be exposed to this (they will only play with `GPUArray`, so didn't pay much attention. But of course I can explicitly default it.

lindsayad commented on framework/include/materials/MaterialPropertyInterface.h:
```diff
@@ -153,6 +158,82 @@ class MaterialPropertyInterface
+#ifdef MOOSE_GPU_SCOPE
+  template <typename T, unsigned int dimension, unsigned int state>
+  GPUMaterialProperty<T, dimension>
+  getGPUGenericMaterialPropertyByName(const std::string & prop_name)
```
multi-line method definitions should go outside the class definition

NamjaeChoi committed:
Change GPU to Kokkos, add namespace, address some minor comments #30655

NamjaeChoi commented on framework/include/gpu/base/GPUMap.h:
```diff
@@ -0,0 +1,84 @@
+  void copy()
+  {
+    std::vector<T1> keys;
+    std::vector<T2> values;
```
Changed.

NamjaeChoi commented on framework/include/gpu/base/GPUVariable.h:
```diff
@@ -0,0 +1,55 @@
+
+class GPUVariable
+{
+  friend class Coupleable;
```
See `GPUCoupleable.K`.

NamjaeChoi commented:
`GPU`s were changed to `Kokkos`s and `Moose::Kokkos` namespace was added. Haven't worked on file/folder names yet. I'm not sure how changing them will affect your review histories, so I will take care of those at the last stage. Docstrings, moving definitions outside, and scope things will be worked soon.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2979828/), step Clang format on 5f090d3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a587c94e57e8343f32b3fde1a800a1561dba71f6`

NamjaeChoi committed:
Check early the existence of Kokkos objects added by actions and initialize Kokkos mesh conditionally #30655

NamjaeChoi committed:
Use private parameter and 'currently executing actions' flag instead of initialize flag in MOOSE classes #30655

NamjaeChoi commented:
My algorithm to sort `_all_objects` storages in `MaterialWarehouse` seems to cause some infinite loops for a certain test problems. Looking for a better algorithm.

NamjaeChoi committed:
Only sort all objects for Kokkos material warehouse #30655

NamjaeChoi commented:
> Do you want me to review this now? Since I usually have many reviews I have to do I typically only look at something when it's non-draft. But I will look at this now if you want me to!

I'm trying to update the code to comply with your programming practices now. Also adding more docstrings. You may be able to start reviewing after that.

lindsayad commented:
Ok I'll unsubscribe for now. Just mention me when it's ready for review

NamjaeChoi committed:
Enhance docstrings, reorganize accessors, move function bodies outside for base classes #30655

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3000566/), step Clang format on 1e5002e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 98f5d7bf5842f4cb14a1dc4be83983680b5c1024`

NamjaeChoi committed:
Enhance docstrings, reorganize accessors, move function bodies outside for kernel classes #30655

NamjaeChoi committed:
Enhance docstrings, reorganize accessors, move function bodies outside for material classes #30655

moosebuild commented:
Job [Controlled apps](https://civet.inl.gov/job/3009841/) on 530350a : invalidated by @loganharbour

NamjaeChoi committed:
Use pass-by-value for performance #30655

lindsayad reviewed (changes requested):
Need documentation. To start, need to describe how to build MOOSE with Kokkos support

NamjaeChoi committed:
Do not allow instantiating vector and matrix outside Kokkos scope #30655

NamjaeChoi committed:
Fix node mapping in multi-process parallelization #30655

NamjaeChoi committed:
Parallel memory pool object #3065

NamjaeChoi committed:
Remove the use of MAX_DOF in assembly #30655

NamjaeChoi committed:
Add configure option for Kokkos #30655

NamjaeChoi committed:
Back to caching elemental data #30655

NamjaeChoi committed:
Use unordered containers for faster mesh update #30655

NamjaeChoi committed:
Reduce local fixed array sizes #30655

NamjaeChoi committed:
Remove support for embedded Kokkos using Makefile due to deprecation #30655

NamjaeChoi committed:
Add install documentations #30655

NamjaeChoi commented on framework/src/gpu/base/GPUMooseInit.K:
```diff
@@ -0,0 +1,50 @@
+{
+  unsigned int num_gpus = Kokkos::num_devices();
+
+  bool has_gpus = comm().verify(num_gpus > 0);
```
Maybe `num_gpus > 0 && comm().verify(num_gpus > 0)`?

loganharbour replied:
This would be a bad race condition if num_gpus wasn’t consistent across all procs

loganharbour replied:
if `num_gpus > 0` is not consistent across all processors, some processors will call MPI and some won't. the processors that call MPI will be waiting for those that didn't, and it'll hang

NamjaeChoi replied:
Okay I see what you mean. Then `comm().verify(num_gpus > 0) && num_gpus > 0`

NamjaeChoi replied:
That also works.

NamjaeChoi commented on framework/src/gpu/systems/GPUNonlinearSystemBase.K:
```diff
@@ -0,0 +1,342 @@
+  else if (ibc)
+  {
+    _kokkos_integrated_bcs.addObject(ibc);
+    _vars[0].addBoundaryVars(boundary_ids, ibc->getCoupledVars());
```
Let's see if any threading test fails for GPU tests after @loganharbour merges my CIVET recipe for Kokkos.

lindsayad replied:
yea I guess there may still be threaded code not covered by your GPU code 👍 yea flagging elements for adaptivity is an example of something I'm sure you haven't hit yet ... although with your productivity I can never be too sure

NamjaeChoi committed:
Address comments #30655

lindsayad committed:
Add compile_commands support for Kokkos sources #30655

lindsayad committed:
Need to perform all device restores before mods on host #30655

NamjaeChoi committed:
Substitute -Werror with -Werror=all-warnings #30655

NamjaeChoi committed:
Make kernels work regardless of MAX_DOF #30655

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3036766/), step Clang format on 0f7e0d7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 178d6cbdfe6ca8c6aba654ed2b3c90b71297ed03`

lindsayad commented:
ah got ya

NamjaeChoi committed:
Add device option to Kokkos tests #30655

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3045192/), step Clang format on f5af724 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b57d1f86f40635e2248990e9de09151eaf643b4a`

NamjaeChoi committed:
Fix creating multiple matrices #30655

NamjaeChoi committed:
Add Kokkos in test harness #30655

NamjaeChoi commented:
I'm working on fixing recover tests now. After that, I won't be updating the code anymore except addressing comments and will be focusing on documentations. If you are waiting for me to stop updating the code for review, now it's the time.

lindsayad commented:
It's a priority for me only behind getting the mfem cuda testing working

NamjaeChoi committed:
Fix stateful restart and recover #30655

NamjaeChoi committed:
Support disabling GPU capabilities for test coverage and debugging purposes #30655

NamjaeChoi committed:
Make code Valgrind-clean #30655

NamjaeChoi committed:
Make code GCC- and clang-clean #30655

NamjaeChoi committed:
Change -imacros to -include for nvcc compatibility #30916

NamjaeChoi committed:
Use libMesh predefined integer types #30655

NamjaeChoi committed:
Change prefixes to Kokkos in tests #30655

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3069868/), step Clang format on d4f112e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f45d794021daae7c06f0977c0b6b2cb772069c29`

lindsayad reviewed (changes requested):
I'll let you know when I'm totally done with review. Will post intermediately to make sure github doesn't lose my comments

lindsayad commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+{
+  destroy();
+
+  _counter = array._counter;
```
Or I guess we'd want `shared_ptr` now? That would make sense

lindsayad commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+               dimension,
+               ").");
+
+  for (unsigned int i = 0; i < d.size(); ++i)
```
```suggestion
  for (const auto i : index_range(d))
```

lindsayad commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+    // If device side memory is not allocated,
+    if (!_is_device_alloc)
+    {
+      if (*_counter == 1)
```
The aliasing stuff makes me nervous because I've already had the concept bite me in MFEM. How often are we aliasing? Is there any reason to consider reusing code in MFEM? This class in particular has many parallels to https://github.com/mfem/mfem/blob/master/general/mem_manager.hpp. I'm guessing it would be impossible to use that given the underlying Kokkos use here

lindsayad commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+  }
+}
+
+template <typename T>
+void
+copyInner(T & /* data */)
+{
+}
+
+template <typename T, unsigned int dimension>
+void
+copyInner(Array<T, dimension> & data)
+{
+  data.copyNested();
+}
```
why these globals instead of just calling class methods directly?

NamjaeChoi replied:
`copyNested` handles copies of nested arrays like `Array<Array<Array<int>>>`. If I make it as a class member, I'm not sure how I should stop recursion at the innermost array.

NamjaeChoi commented:
> I'll let you know when I'm totally done with review. Will post intermediately to make sure github doesn't lose my comments

Okay thank you. I will address them at once.

NamjaeChoi commented on framework/app.mk:
```diff
@@ -593,3 +619,4 @@ endif
 sa: $(app_analyzer)
 
 compile_commands_all_srcfiles += $(srcfiles)
+compile_commands_all_kokkos_srcfiles += $(app_KOKKOS_SRC_FILES)
```
It's still there.

lindsayad replied:
I'm dumb. I was expecting it to be listed next to `app.mk` and `build.mk` but the lexicographical ordering puts directories in the middle

NamjaeChoi committed:
Check nodal definedness in nodal kernels and BCs #30655

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1589 @@
+};
+
+template <typename T>
+class Array<T, 5> : public ArrayBase<T, 5>
```
On the material side, I allow users to go up to 4D material properties, which becomes a 5D array with the last index being global QP.

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1589 @@
+   */
+  KOKKOS_FUNCTION T & operator()(dof_id_signed_type i0) const
+  {
+    KOKKOS_ASSERT(i0 - _d[0] >= 0 && static_cast<dof_id_type>(i0 - _d[0]) < _n[0]);
```
Not sure how dynamic casting will work on GPU. That static cast is just to silence compilers complaining about signed and unsigned comparison.

NamjaeChoi committed:
Fix Kokkos dependency treatment in makefile #30655

loganharbour reviewed (changes requested):
I'm at 75/300.. so that's good

NamjaeChoi committed:
Attempt to build separate libraries #30655

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * Function pointer maps for load/store
+   */
+  ///{@
+  static std::unordered_map<PropertyKey, PropertyStore> _store;
+  static std::unordered_map<PropertyKey, PropertyLoad> _load;
```
How is it related with multiapps?

loganharbour replied:
This map will be shared across multiapps. Which means if that property key is not unique across multiapps, you will have collisions

NamjaeChoi replied:
The purpose of this is to only instantiate `dataLoad` and `dataStore` for the data types that are used for stateful properties. The key is of type `std::pair<std::type_index, unsigned int>`, where `std::type_index` is for the data type of material property and `unsigned int` is for the property dimension. This function pointer map stores the instantiations of the following template functions in the same source file:

```
template <typename T, unsigned int dimension>
void
propertyStore(std::ostream & stream, void * prop)
{
  auto property = static_cast<MaterialProperty<T, dimension> *>(prop);

  dataStore(stream, property->_data, nullptr);
}
template <typename T, unsigned int dimension>
void
propertyLoad(std::istream & stream, void * prop)
{
  auto property = static_cast<MaterialProperty<T, dimension> *>(prop);

  dataLoad(stream, property->_data, nullptr);
}
```

`property->_data` is of `Moose::Kokkos::Array<T, dimension + 1>` type.

Key being not unique across multiapps means the properties are of the same type, and thus they can share the same function.

loganharbour replied:
Is there not another place that we can put the state? In general, we've really only used static variables for const constants and for registries. It opens up a potential can of worms in the future when we could have multiple instantiations of an app writing to the same structure

NamjaeChoi replied:
I will try moving it to the material property storage.

NamjaeChoi replied:
The test object illustrating this situation is `GPUNontrivialStateful.K`

lindsayad replied:
That makes sense. I like the static maps myself. @loganharbour they basically are registries with one entry per template. It has nothing to do with the number of instances of a given template. As long as `registerLoadStore` is always done non-threaded, then I don't think there are any safety concerns

NamjaeChoi replied:
I pushed this in a separate commit so that we can revert it later depending on our decision.

NamjaeChoi commented on framework/include/gpu/bcs/GPUCoupledVarNeumannBC.h:
```diff
@@ -0,0 +1,56 @@
+      return 0;
+  }
+
+protected:
```
The reason for this is described in the class docstring of `ReferenceWrapper`. The basic reason is because reference refers to a host variable and thus cannot be accessed on device. All GPU objects that are to be accessed should not be stored as reference. Adding const should be fine I guess.

NamjaeChoi commented on framework/include/gpu/kernels/GPUCoupledForce.h:
```diff
@@ -0,0 +1,39 @@
+ * Implements a source term proportional to the value of a coupled variable. Weak form: $(\\psi_i,
+ * -\\sigma v)$.
+ */
+class KokkosCoupledForce final : public Moose::Kokkos::Kernel<KokkosCoupledForce>
```
This will be handled all at once in order not to screw up your review histories, so leaving it for now.

NamjaeChoi commented on framework/include/gpu/systems/GPUVector.h:
```diff
@@ -0,0 +1,199 @@
+   * @param system The Kokkos system
+   * @param assemble Whether the vector will be assembled
+   */
+  void create(libMesh::NumericVector<PetscScalar> & vector, const System & system, bool assemble);
```
No it can't be. I need to obtain a writeable PETSc vector.

NamjaeChoi commented on framework/include/base/MooseApp.h:
```diff
@@ -368,6 +378,11 @@ class MooseApp : public ConsoleStreamInterface,
+  /**
+   * Returns whether the simulation is currently executing actions
+   */
+  bool currentlyExecutingActions() const { return _currently_executing_actions; }
```
You mean check whether `ActionWarehouse::getCurrentAction()` returns null or not?

loganharbour replied:
Yep, exactly. Just trying to avoid keeping around extra state that we forget to update.

NamjaeChoi commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -64,6 +73,7 @@ class TaggingInterface
     friend class LinearSystemContributionObject;
     template <typename>
     friend class MooseObjectTagWarehouse;
+    friend class Moose::Kokkos::ResidualObject;
```
I need to get the vector/matrix tags that the residual object should fill.

NamjaeChoi commented on framework/include/gpu/mesh/GPUMesh.h:
```diff
@@ -0,0 +1,341 @@
+   * @param elem The libMesh element
+   * @returns The element type ID
+   */
+  unsigned int getElementTypeID(const Elem * elem) const;
+  /**
+   * Get the serialized element ID of an element
+   * @param elem The libMesh element
+   * @returns The element ID
+   */
+  dof_id_type getElementID(const Elem * elem) const;
+  /**
+   * Get the serialized subdomain-local element ID of an element
+   * The ID starts from zero in each subdomain
+   * @param elem The libMesh element
+   * @returns The subdomain-local element ID
+   */
+  dof_id_type getSubdomainLocalElementID(const Elem * elem) const;
```
No particular reason, just because many MOOSE things receive or return Elem pointer.

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+ */
+enum class MemcpyKind
+{
+  HOST_TO_DEVICE,
+  DEVICE_TO_HOST,
+  DEVICE_TO_DEVICE
```
Do we?

NamjaeChoi commented on framework/include/base/MooseApp.h:
```diff
@@ -1565,6 +1583,71 @@ class MooseApp : public ConsoleStreamInterface,
   friend class Restartable;
   friend class SubProblem;
+
+public:
```
Should I rebase?

NamjaeChoi commented on framework/include/base/Assembly.h:
```diff
@@ -223,6 +223,15 @@ class Assembly
+   * Returns the reference to the quadrature of specified block and dimension
+   * @return A pointer.
+   */
+  libMesh::QBase * writeableQRule(unsigned int dim, SubdomainID block)
```
How about the existing one?

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+/**
+ * A structure storing the metadata of Kokkos material properties
+ */
+struct PropRecord
```
But it's inside namespace? I will leave this until we decide how we are gonna use namespace.

lindsayad replied:
What namespace? If it's in `Moose` you don't need qualification in `Moose::Kokkos`. I'm still all aboard the `Moose::Kokkos` train

NamjaeChoi replied:
I thought Logan was telling me to change it to some other name,

NamjaeChoi committed:
Addressing comments #30655

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+  /**
+   * Reference counter
+   */
+  unsigned int * _counter = nullptr;
```
Using `shared_ptr`

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+template <typename T, unsigned int dimension = 1>
+class Array
+{
+};
```
I had a reason that no longer exists.

NamjaeChoi committed:
Increase coverage #30655

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1601 @@
+   * @returns The reference of the entry depending on the architecture this function is being
+   * called on
+   */
+  KOKKOS_FUNCTION T & operator()(dof_id_signed_type i0) const
```
No strong motivation. Simply to be consistent with multi-D operators?

lindsayad replied:
that sounds good

NamjaeChoi replied:
And actually there is one more thing. `operator[]` is currently being used for dimensionless accessor.

lindsayad commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+ */
+class MaterialPropertyBase
+{
+  friend class MaterialPropertyStorage;
```
Move to bottom

loganharbour commented on framework/include/gpu/base/GPUAssembly.h:
```diff
@@ -0,0 +1,519 @@
+
+#ifdef MOOSE_KOKKOS_SCOPE
+KOKKOS_FUNCTION inline Real
+Assembly::coordTransformFactor(SubdomainID subdomain, Real3 point) const
```
Sounds good. Is there a reason to not use const?

NamjaeChoi replied:
It was changed to const.

loganharbour commented on framework/include/base/Assembly.h:
```diff
@@ -211,6 +219,19 @@ class Assembly
+   * classes may create keys using simple curly braces \p {} or may be more explicit and use \p
+   * Assembly::InternalDataKey{}
+   */
+  class InternalDataKey
```
There's a `MoosePassKey` helper for this.

```
using InternalDataKey = MoosePassKey<Moose::Kokkos::Assembly>;
```

loganharbour commented on framework/include/gpu/kernels/GPUKernelGrad.h:
```diff
@@ -0,0 +1,196 @@
+KOKKOS_FUNCTION void
+KernelGrad<Derived>::computeResidualInternal(const Derived * kernel, ResidualDatum & datum) const
+{
+  Real local_re[MAX_DOF];
+
+  unsigned int num_batches = datum.n_dofs() / MAX_DOF;
+
+  if (datum.n_dofs() % MAX_DOF)
+    ++num_batches;
+
+  for (unsigned int batch = 0; batch < num_batches; ++batch)
+  {
+    unsigned int ib = batch * MAX_DOF;
+    unsigned int ie = ::Kokkos::min(ib + MAX_DOF, datum.n_dofs());
+
+    for (unsigned int i = ib; i < ie; ++i)
+      local_re[i - ib] = 0;
+
+    for (unsigned int qp = 0; qp < datum.n_qps(); ++qp)
+    {
+      datum.reinit();
+
+      Real3 value = datum.JxW(qp) * kernel->precomputeQpResidual(qp, datum);
+
+      for (unsigned int i = ib; i < ie; ++i)
+        local_re[i - ib] += value * _grad_test(datum, i, qp);
+    }
+
+    for (unsigned int i = ib; i < ie; ++i)
+      accumulateTaggedElementalResidual(local_re[i - ib], datum.elem().id, i);
+  }
```
This logic seems pretty common across a few methods and kernels - is there a possibility to put it into a helper where you can pass in a lambda? Since it's all in the header, I suspect the compiler would be smart enough to inline everything

NamjaeChoi replied:
I also felt that but not sure how. I will think about this more.

NamjaeChoi replied:
Pushed this in a separate commit. Lambda seems to work fine with CUDA. Hopefully it works for others too.

NamjaeChoi replied:
The Jacobian loop in `GPUTimeDerivative` was left there because it has the mass matrix lumping option that can't be commonly factored out. This can be an example for developers/users how to customize those loops.

loganharbour commented on framework/include/gpu/materials/GPUMaterialBase.h:
```diff
@@ -0,0 +1,213 @@
+   * Kokkos function tags
+   */
+  ///@{
+  struct ElementInit
```
Is there a reason why these aren't just enums?

NamjaeChoi replied:
It's a way how Kokkos overloads `operator()`

loganharbour commented on framework/include/interfaces/BoundaryRestrictable.h:
```diff
@@ -220,12 +233,59 @@ class BoundaryRestrictable
    */
   bool hasBoundaryMaterialPropertyHelper(const std::string & prop_name) const;
+
+#ifdef MOOSE_KOKKOS_SCOPE
```
I thought I commented on this already, but maybe not. These methods are only used for kokkos - should they have a name that describes them as such?

loganharbour commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -330,6 +331,22 @@ class MaterialPropertyStorage
+  /**
+   * @return The consumers of storage of type \p type
+   */
+  const std::set<const MooseObject *> & getConsumers(Moose::MaterialDataType type)
```
```suggestion
  const std::set<const MooseObject *> & getConsumers(Moose::MaterialDataType type) const
```

loganharbour commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -370,6 +390,9 @@ class MaterialPropertyStorage
   std::vector<unsigned int> _stateful_prop_id_to_prop_id;
 
+  /// The consumers of this storage
+  std::map<Moose::MaterialDataType, std::set<const MooseObject *>> _consumers;
```
How did this not exist before? I'm a bit confused

loganharbour commented on framework/include/materials/MaterialWarehouse.h:
```diff
@@ -42,7 +42,7 @@ class MaterialWarehouse : public MooseObjectWarehouse<MaterialBase>
   virtual void residualSetup(THREAD_ID tid = 0) const;
   virtual void updateActive(THREAD_ID tid = 0);
-  void sort(THREAD_ID tid = 0);
+  void sort(THREAD_ID tid = 0, bool sort_all_objects = false);
```
Can you please add a docstring?

It's weird seeing a method called `sort` that doesn't by default sort everything? I understand this is the current behavior, but it's still odd

NamjaeChoi replied:
In case you haven't noticed:
```
/**
   * By default, this method only sorts block and boundary-wise object storages that are used by the
   * MOOSE threaded element loops. Kokkos, however, computes all elements and faces at once
   * regardless of block and boundary and uses all-object storages. Therefore, the Kokkos material
   * warehouse sets \p sort_all_objects to true to sort the all-object storages.
   */
```

loganharbour commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -390,6 +408,10 @@ class FEProblemBase : public SubProblem, public Restartable
   virtual void init() override;
   virtual void solve(const unsigned int nl_sys_num);
 
+#ifdef MOOSE_KOKKOS_ENABLED
```
Should add a docstring that summarizes that happens here

loganharbour commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2464,6 +2553,8 @@ class FEProblemBase : public SubProblem, public Restartable
 
   virtual Moose::FEBackend feBackend() const { return Moose::FEBackend::LibMesh; }
 
+  bool hasKokkosObjects() { return _has_kokkos_objects; }
```
```suggestion
  bool hasKokkosObjects() const { return _has_kokkos_objects; }
```

and add a docstring

loganharbour commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -856,6 +920,10 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
   MooseObjectWarehouse<DirichletBCBase> _preset_nodal_bcs;
   MooseObjectWarehouse<ADDirichletBCBase> _ad_preset_nodal_bcs;
+
+  MooseObjectTagWarehouse<ResidualObject> _kokkos_integrated_bcs;
```
All of these new warehouses should be ifdef-ed on kokkos, yeah?

loganharbour commented on framework/src/base/Moose.C:
```diff
@@ -519,7 +519,9 @@ associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
   registerSyntaxTask("CopyNodalVarsAction", "AuxVariables/*", "copy_nodal_aux_vars");
 
   registerSyntaxTask("AddKernelAction", "Kernels/*", "add_kernel");
+  registerSyntaxTask("AddKokkosKernelAction", "KokkosKernels/*", "add_kernel");
```
These tasks should be ifdef-ed on `MOOSE_KOKKOS_ENABLED`?

NamjaeChoi replied:
These object actions will be registered but not the actual objects. Let me try what happens exactly when I try to run a Kokkos input with non-Kokkos MOOSE.

NamjaeChoi replied:
That means we should always compile `*.K` files? That will be very difficult.

NamjaeChoi replied:
I will choose to disable the syntaxes completely when Kokkos is disabled.

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -863,6 +869,28 @@ MooseApp::registerCapabilities()
+#else
+    missingCapability("kokkos",
+                      doc,
+                      "Rebuild PETSc with Kokkos support and libMesh, or checkout the Kokkos "
+                      "submodule. Then, rebuild MOOSE with KOKKOS=true.");
```
We don't have a kokkos submodule. Are these instructions old?

NamjaeChoi replied:
Good catch

loganharbour commented on framework/src/base/MooseObject.C:
```diff
@@ -52,6 +52,11 @@ MooseObject::MooseObject(const InputParameters & parameters)
 {
+  // Calling this constructor while not executing actions means this object is being
+  // copy-constructed
+  if (parameters.isParamValid("_kokkos_object") && !_app.currentlyExecutingActions())
+    return;
```
Don't just return here in case we poorly add something after the error check that might get skipped.

Instead of checking `_kokkos_object` everywhere, we should create a static var with this name. Something like:

```
namespace Moose::Kokkos
{
  static const kokkos_object_param = "_kokkos_object";
}
```

where we then just use`Moose::Kokkos::kokkos_object_param` instead. And add a member function to `MooseObject` (when kokkos is enabled), with

```
#ifdef MOOSE_KOKKOS_ENABLED
bool isKokkosObject() const return {parameters().isParamValid(Moose::Kokkos::kokkos_object_Param);
#endif
```

which you then check here

NamjaeChoi replied:
Errors should be hit during their construction in actions not during copy, right?

loganharbour replied:
Correct - what are you thinking?

NamjaeChoi replied:
The return happens when this constructor is being invoked during copy construction (functor dispatch). Only happens after actions.

loganharbour commented on framework/src/base/ResidualObject.C:
```diff
@@ -54,6 +54,11 @@ ResidualObject::ResidualObject(const InputParameters & parameters, bool is_nodal
 {
+  // Calling this constructor while not executing actions means this object is being
+  // copy-constructed
+  if (parameters.isParamValid("_kokkos_object") &&
```
Does this even do anything? Same with the rest like this

loganharbour commented on framework/src/gpu/actions/AddGPUMaterialAction.C:
```diff
@@ -0,0 +1,35 @@
+{
+  if (!_moose_object_pars.get<bool>("_interface"))
+    _problem->addKokkosMaterial(_type, _name, _moose_object_pars);
+  // else
+  //  _problem->addKokkosInterfaceMaterial(_type, _name, _moose_object_pars);
```
?

loganharbour replied:
We typically don't leave commented code in like this, more so when it's something very simple

NamjaeChoi replied:
Deleted

loganharbour reviewed (commented):
Up to 105 files reviewed

loganharbour reviewed (changes requested):
Up to 133 files. Getting there... off for a bit.

NamjaeChoi commented on framework/src/gpu/bcs/KokkosConvectiveFluxBC.K:
```diff
@@ -0,0 +1,34 @@
+  params.set<Real>("initial") = 500;
+  params.set<Real>("final") = 500;
+  params.set<Real>("duration") = 0.0;
+  params.addClassDescription(
+      "Determines boundary values via the initial and final values, flux, and exposure duration");
```
Well... this was just copied from MOOSE `ConvectiveFluxBC`

NamjaeChoi replied:
Moved to test folder

NamjaeChoi committed:
Increase coverage 2 #30655

NamjaeChoi committed:
Address comments 2 #30655

NamjaeChoi commented on framework/src/gpu/kernels/GPUReaction.K:
```diff
@@ -0,0 +1,29 @@
+      "Implements a simple consuming reaction term with weak form $(\\psi_i, \\lambda u_h)$.");
+  params.addParam<Real>(
+      "rate", 1.0, "The $(\\lambda)$ multiplier, the relative amount consumed per unit time.");
+  params.declareControllable("rate");
```
See `kokkos_2d_diffusion_bodyforce_test.i`. It has a.changing postprocessor value over time.

NamjaeChoi replied:
I guess your concern is it doesn't hold reference properly.

NamjaeChoi replied:
It holds a reference to the host data. It also holds a value to be accessed on device. The value is copied from the host reference in the copy constructor.

NamjaeChoi replied:
See `ReferenceWrapper` which is the base class

loganharbour commented on framework/include/gpu/base/GPUVariable.h:
```diff
@@ -0,0 +1,133 @@
+ */
+class Variable
+{
+  friend class ::Coupleable;
```
Can we not only expose a few things with the passkey idiom?

NamjaeChoi replied:
Design changed

loganharbour reviewed (changes requested):
Down to the test objects now.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3078250/), step Verify coverage on ca64c49 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

NamjaeChoi commented on modules/heat_transfer/src/gpu/kernels/GPUHeatConduction.K:
```diff
@@ -0,0 +1,38 @@
+      "thermal_conductivity_dT",
+      "Property name of the derivative of the diffusivity with respect "
+      "to the variable (Default: thermal_conductivity_dT)");
+  // params.set<bool>("use_displaced_mesh") = true;
```
Not supported yet

NamjaeChoi replied:
Deleted

NamjaeChoi commented on framework/include/gpu/nodalkernels/KokkosUpperBoundNodalKernel.h:
```diff
@@ -0,0 +1,79 @@
+  const Real _upper_bound;
+};
+
+KOKKOS_FUNCTION inline Real
+KokkosUpperBoundNodalKernel::computeQpResidual(const dof_id_type node) const
+{
+  for (dof_id_type b = 0; b < _bnd_ids.size(); ++b)
+    if (kokkosMesh().isBoundaryNode(node, _bnd_ids[b]))
+      return _u(node);
+
+  return ::Kokkos::min(_u(node), _upper_bound - _v(node));
+}
```
This one I can factor out with minimal impact.

NamjaeChoi replied:
You mean I should make an intermediate base class?

NamjaeChoi commented on framework/include/gpu/kernels/GPUBodyForce.h:
```diff
@@ -0,0 +1,58 @@
+    params.addParam<Real>("value", 1.0, "Coefficient to multiply by the body force term");
+    params.addParam<PostprocessorName>(
+        "postprocessor", 1, "A postprocessor whose value is multiplied by the body force");
+    params.declareControllable("value");
```
If you are convinced with `ReferenceWrapper`, please resolve this one and the following one too.

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialPropertyStorage.h:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Record of each material property
+   */
+  std::map<std::string, Moose::Kokkos::PropRecord> _kokkos_prop_records;
```
Changed to unordered_map. Is this discussion closed?

lindsayad replied:
Let's assume that the map is large and the strings are short. In that case `unordered_map` is the winner. If we something bad in profiling we can change

NamjaeChoi commented on framework/src/interfaces/BlockRestrictable.C:
```diff
@@ -84,7 +86,11 @@ BlockRestrictable::initializeBlockRestrictable(const MooseObject * moose_object)
+        &_blk_feproblem->getMaterialData(Moose::BLOCK_MATERIAL_DATA,
+                                         _blk_tid,
+                                         nullptr,
+                                         moose_object->isParamValid("_kokkos_object"));
```
How about passing in moose object and let material data query

NamjaeChoi replied:
Can't think of a good idea. If MOOSE object is passed in, it thinks that the object is a consumer currently. Unless we change this logic, it's not possible to determine which material data should be returned without explicit argument.

NamjaeChoi replied:
I just separated `getKokkosMaterialData`

lindsayad commented on framework/src/gpu/kernels/GPUMatCoupledForce.K:
```diff
@@ -0,0 +1,67 @@
+    _coef(isParamValid("coef") ? getParam<std::vector<Real>>("coef")
+                               : std::vector<Real>(_n_coupled, 1))
+{
+  for (MooseIndex(_n_coupled) j = 0; j < _n_coupled; ++j)
```
```suggestion
  for (const auto j : make_range(_n_coupled))
```
We prefer `make_range`/`index_range` over `MooseIndex` these days

lindsayad commented on framework/src/gpu/kernels/GPUMatCoupledForce.K:
```diff
@@ -0,0 +1,67 @@
+  if (_coupled_props)
+  {
+    _mat_props.create(_n_coupled);
+    std::vector<MaterialPropertyName> names =
```
```suggestion
    const auto & names =
```

lindsayad commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * Get whether this property is valid
+   * @returns Whether this property is valid
+   */
+  KOKKOS_FUNCTION operator bool() const { return _id != libMesh::invalid_uint || _default; }
```
I would prefer a more explicit method like `isValid()` or something

NamjaeChoi replied:
I can add `isValid()`. This mimics the behavior of optional material property.

lindsayad replied:
Hmm if there's already precedence it's fine

lindsayad commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * @param qp The local quadrature point index
+   * @returns The MaterialPropertyValue object that provides access to the property values
+   */
+  KOKKOS_FUNCTION MaterialPropertyValue<T, dimension> operator()(Datum & datum,
```
Where is this defined?

NamjaeChoi replied:
`GPUMaterialPropertyValue.h`

lindsayad replied:
Why isn't it in `GPUMaterialProperty.h`? Methods for `Moose::Kokkos::MaterialProperty` should be in the same header as the class definition

NamjaeChoi replied:
It's a bit entangled. It creates a `MaterialPropertyValue` object so it should see its definition, but `MaterialPropertyValue` also needs to see `MaterialProperty`

lindsayad replied:
This is how we would handle this in MetaPhysicL (where we do this kind of thing all the time):
- `KokkosMaterialPropertyValue_decl.h` - class definition of `Moose::Kokkos::MaterialPropertyValue`. Should be able to operate with just a forward declaration of `Moose::Kokkos::MaterialProperty` 
- `KokkosMaterialProperty_decl.h` - class definition of `Moose::Kokkos::MaterialProperty`. Should be able to operate with just a forward declaration of `Moose::Kokkos::MaterialPropertyValue`
- The existing headers you have would remain and include method definitions and include both of the above `_decl` headers

NamjaeChoi replied:
Actually NVM, I think I can get around this.

NamjaeChoi replied:
Resolved

lindsayad commented:
> What the heck is this?

This will happen sometimes if the base shas end up being different for different coverage recipes
```
>> Loading head coverage reports
     id    job_id    recipe_id  tag                           head_sha                                  base_sha
-------  --------  -----------  ----------------------------  ----------------------------------------  ----------------------------------------
1173827   3078234        47583  framework-torch               ca64c493873c45ad014719150f14d70124fc70fe  d5af4c557e77154491814ec610786411396b06b0
1173902   3078231        48068  framework2                    ca64c493873c45ad014719150f14d70124fc70fe  d5af4c557e77154491814ec610786411396b06b0
1173897   3078232        47537  framework_increased_sparsity  ca64c493873c45ad014719150f14d70124fc70fe  d5af4c557e77154491814ec610786411396b06b0
1173939   3078230        48090  framework1                    ca64c493873c45ad014719150f14d70124fc70fe  d5af4c557e77154491814ec610786411396b06b0
1174197   3078228        47851  framework-cuda-gpu            ca64c493873c45ad014719150f14d70124fc70fe  2ad78df8c544c9cf2cf0b743b06b5557ce23a663
```

NamjaeChoi commented on framework/include/gpu/bcs/GPUNodalBC.h:
```diff
@@ -0,0 +1,224 @@
+{
+
+/**
+ * The base class for a user to derive his own Kokkos nodal boundary conditions.
```
This is just an English question. Is `their` correct here?

NamjaeChoi replied:
Thank you, teacher 😄

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialPropertyStorage.h:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Record of each material property
+   */
+  std::unordered_map<std::string, Moose::Kokkos::PropRecord> _kokkos_prop_records;
```
This is due to ambiguity, because its parent class (MOOSE `MaterialPropertyStorage`) has `PropRecord` definition.

lindsayad replied:
Ah I see because it's nested within the class

lindsayad reviewed (changes requested):
Finished the material section

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * Get the clone of this property
+   * @returns The clone this property constructed through the copy constructor
+   */
+  MaterialProperty<T, dimension> clone() { return MaterialProperty<T, dimension>(*this); }
```
These are not equivalent. The changed version will trigger C++ copy elision and does not call copy constructor.

lindsayad replied:
If copy elision vs. copy construction matters in the class, the class design is not value-semantic, and that is the real concern

lindsayad replied:
In the current design, the compiler may still elide construction

NamjaeChoi replied:
I will think over the design then. In fact this should not be called by users. Material properties cloned from the reference material property are value-semantic, but reference material properties requires caution.

NamjaeChoi replied:
This was addressed (IMO). Now copy construction vs copy elision doesn't matter. Please take a look.

NamjaeChoi commented on framework/include/gpu/mesh/GPUMesh.h:
```diff
@@ -0,0 +1,379 @@
+   */
+  void update();
+
+  /**
+   * Get the serialized subdomain ID of a subdomain
+   * @param subdomain The MOOSE subdomain ID
+   * @returns The subdomain ID
+   */
+  SubdomainID getSubdomainID(const SubdomainID subdomain) const;
+  /**
+   * Get the serialized boundary ID of a boundary
+   * @param boundary The MOOSE boundary ID
+   * @returns The boundary ID
+   */
+  BoundaryID getBoundaryID(const BoundaryID boundary) const;
```
The reason is simple. It is because I need IDs that start from zero and are sequential. Maps are expensive on GPU.

lindsayad replied:
I think we should change the names then. `getContiguousXXXID` maybe

NamjaeChoi replied:
Serialized?

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+  virtual void swap(MaterialPropertyBase & prop) override;
+#endif
+
+  /**
+   * Pointer to the reference property
+   */
+  const MaterialProperty<T, dimension> * _reference = nullptr;
```
The `_data` arrays are not allocated until `FEProblemBase::initKokkos()`, which is well after all the objects acquired their material properties during construction. But because of this synchronization, it works fine.

NamjaeChoi replied:
This was included in the doxygen

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterial.h:
```diff
@@ -0,0 +1,459 @@
+  virtual Moose::MaterialDataType materialDataType() override { return _material_data_type; }
+
+  /**
+   * Flag whether the material is on faces
+   */
+  const bool _bnd;
```
This also mimics MOOSE materials.

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * Constructor for default property
+   * @param value The default value
+   */
+  MaterialProperty(const T value)
```
I just can't guess how heavy `T` can get. It can be any struct as far as it doesn't contain dynamic allocation.

lindsayad replied:
Let's use references then

NamjaeChoi replied:
Done

lindsayad commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+namespace Kokkos
+{
+
+void free(void * ptr);
```
why the rename?

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialPropertyStorage.h:
```diff
@@ -0,0 +1,144 @@
+   */
+  static MaterialPropertyStorage & cast(::MaterialPropertyStorage & storage)
+  {
+    return static_cast<MaterialPropertyStorage &>(storage);
```
I was confused by some compiler errors while attempting this and realized that the there is no virtual override in the inheritance. So I can't test dynamic cast here.

lindsayad replied:
Ah that makes sense. Sorry for the time wasted

NamjaeChoi commented on framework/include/gpu/base/GPUReferenceWrapper.h:
```diff
@@ -0,0 +1,264 @@
+  const T _copy;
+};
+
+// The const specialization
+template <typename T>
+class ReferenceWrapper<const T>
```
That's surprising, because in the non-const class I have both const and non-const methods. Don't they become ambiguous when T is a const type?

lindsayad replied:
No because what method is called is determined by the const-ness of the `ReferenceWrapper` itself

NamjaeChoi replied:
Deleted

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterial.h:
```diff
@@ -0,0 +1,459 @@
+Material<Derived>::operator()(NeighborInit, const dof_id_type tid) const
+{
+  auto material = static_cast<const Derived *>(this);
+  auto elem = kokkosElementSideID(tid);
```
Actually, `NeighborInit` is a kind of dummy code now. It will have a check on the neighbor existence later.

NamjaeChoi replied:
I guess structured binding will be doable. Let me check

NamjaeChoi committed:
Addressing comments 3 #30655

NamjaeChoi committed:
Make load/store map non-static #30655

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+  /**
+   * Default value
+   */
+  T _value;
```
No because const should be always initialized?

NamjaeChoi replied:
I realized that _value can be copied from another material property. It allows shallow copy.

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+   * Store the data to a stream
+   * @param stream The stream to store
+   */
+  void store(std::ostream & stream) { _store[key()](stream, this); }
```
Static maps were changed to nonstatic maps and are in material property storage now. It's in a separate commit. We can revert it depending on what we choose to do.

NamjaeChoi committed:
Addressing comments 4 #30655

NamjaeChoi committed:
Fix example/tutorial build #30655

NamjaeChoi commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -59,6 +59,12 @@ Coupleable::Coupleable(const MooseObject * moose_object, bool nodal, bool is_fv)
+  // copy-constructed
+  if (moose_object->isParamValid("_kokkos_object") &&
+      !moose_object->getMooseApp().currentlyExecutingActions())
+    return;
```
`isKokkosObject` added

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+    if (prop_cast._data[i].isAlloc())
+      _data[i].deepCopy(prop_cast._data[i]);
+
+  _data.copy();
```
Separated `copyToHost` and `copyToDevice`

NamjaeChoi replied:
Validity checking will be possible if we very conservatively check it (like whenever operator() is called or pointer was obtained on host, we invalidate device). But not sure if we want that level of conservatism.

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialBase.h:
```diff
@@ -0,0 +1,213 @@
+   * Get the material data type
+   * @returns The material data type
+   */
+  virtual Moose::MaterialDataType materialDataType() = 0;
```
Moved

NamjaeChoi commented on framework/include/gpu/base/GPUArray.h:
```diff
@@ -0,0 +1,1529 @@
+   * Copy data between host and device
+   * @param dir The copy direction
+   */
+  void copy(MemcpyKind dir = MemcpyKind::HOST_TO_DEVICE);
```
Separated

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterialBase.h:
```diff
@@ -0,0 +1,213 @@
+#include "GPUTypes.h"
+#include "GPUMaterialPropertyValue.h"
+
+// initQpStatefulProperties() and computeQpProperties() are intentionally hidden
+// but some compilers generate ugly warnings
+
+#if defined(__clang__)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Woverloaded-virtual"
+#elif defined(__GNUC__) || defined(__GNUG__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Woverloaded-virtual"
+#endif
```
This doesn't work well. Virtual override can violate accessors.

NamjaeChoi committed:
Addressing comments 5 #30655

NamjaeChoi commented on framework/include/gpu/materials/GPUMaterial.h:
```diff
@@ -0,0 +1,459 @@
+   */
+  template <typename T, unsigned int dimension, unsigned int state>
+  MaterialProperty<T, dimension>
+  getKokkosGenericMaterialPropertyByName(const std::string & prop_name_in);
```
Changed

NamjaeChoi committed:
Factor out common residual and Jacobian loop #30655

NamjaeChoi committed:
Addressing comments 6 #30655

NamjaeChoi commented:
Do you think I can rebase once, if you don't have pending comments?

lindsayad commented on framework/include/gpu/materials/GPUMaterialProperty.h:
```diff
@@ -0,0 +1,412 @@
+
+  for (auto subdomain : subdomains)
+  {
+    auto sid = mesh.getKokkosMesh()->getSubdomainID(subdomain);
```
we won't drop this line, but we'll change the method name. Marking this comment as resolved

NamjaeChoi committed:
Change prefixes to Kokkos in sources #30655

NamjaeChoi committed:
Addressing comments 7 #30655

NamjaeChoi committed:
Fix restep test #30655

NamjaeChoi committed:
Improve material property design #30655

NamjaeChoi committed:
Resolve material property entanglement #30655

NamjaeChoi committed:
Remove unnecessary inline keyword #30655

NamjaeChoi committed:
Change devices to compute_devices #30655

NamjaeChoi committed:
Add object documentations #30655

NamjaeChoi committed:
Add syntax documentations #30655

lindsayad commented on test/tests/kokkos/kernels/block_kernel/kokkos_block_kernel_test.i:
```diff
@@ -0,0 +1,52 @@
+[]
+
+[Variables]
+  [./u]
```
would be nice to remove use of the old-style syntax

lindsayad reviewed (changes requested):
Does Kokkos MOOSE support adaptivity? If so, would be good to add a test. If not, should add an error check and probably also make an issue

NamjaeChoi committed:
Update old syntax #30655

NamjaeChoi committed:
Improve map implementation #30655

NamjaeChoi committed:
Error on displacement and adaptivity #30655

NamjaeChoi commented:
Removed the use of extra block and boundary IDs by leveraging hash maps for data storage, but depending on the performance, I may revert it.

NamjaeChoi committed:
Remove subdomain local element ID map #30655

NamjaeChoi committed:
Cache reference element data separately #30655

NamjaeChoi committed:
Use type aliases for Kokkos IDs #30655

NamjaeChoi commented:
Using maps results in noteable performance drop and memory increase. I will keep investigating ways to remove the "contiguous" IDs. For now, I'm using type alias for element, node, subdomain, and boundary IDs.

NamjaeChoi committed:
Increase coverage 3 #30655

NamjaeChoi committed:
Update module objects to match up the original versions #30655

NamjaeChoi committed:
Add module tests #30655

NamjaeChoi committed:
Fix SQA and confilicts #30655

lindsayad commented:
I moved to-dos before merging into the OP and marked off a couple items. The only unmarked item is the copy construction. It seems like we should talk about that more since at a recent NEAMS MP meeting you revealed that you don't in fact want (true/deep) copy construction, which I suspected all along 😄 

@loganharbour please add to the to-do list in the OP if you think of other needed to-dos before merge

NamjaeChoi commented:
Still working on syntax documentations, just FYI.

NamjaeChoi committed:
Sanity check for Kokkos array initialization status #30655

NamjaeChoi committed:
Define special passkey-protected constructors for functor copy #30655

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3201703/), step Clang format on f652b6b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30835/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f149782cc7d0a498e664639e75ae48ac7b9a8a02`

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1140,6 +1170,13 @@ FEProblemBase::initialSetup()
         _has_initialized_stateful = true;
+
+#ifdef MOOSE_KOKKOS_ENABLED
+      if (_kokkos_material_props.hasStatefulProperties() ||
```
```suggestion
      else if (_kokkos_material_props.hasStatefulProperties() ||
```

loganharbour commented on framework/src/utils/MooseTypes.C:
```diff
@@ -27,6 +27,15 @@ const TagName OLD_SOLUTION_TAG = "SOLUTION_STATE_1";
+#ifdef MOOSE_KOKKOS_ENABLED
+namespace Kokkos
+{
+const std::string KOKKOS_OBJECT_PARAM = "_kokkos_object";
```
Can you move this to `MooseBase` along with the rest of the parameter names? I think it fits better there

NamjaeChoi replied:
Added in `MooseBase`

NamjaeChoi commented on framework/include/base/MooseObject.h:
```diff
@@ -31,6 +31,13 @@ class MooseObject : public ParallelParamObject, public std::enable_shared_from_t
+  /**
+   * Special constructor used for Kokkos functor copy during parallel dispatch
+   */
+  MooseObject(const MooseObject & object, const Moose::Kokkos::FunctorCopy &);
```
I can remove reference in most places including this, but it is required in some places because the key cannot be copied around.

loganharbour replied:
Isn't the purpose of the key to not be copied?

loganharbour replied:
It really doesn't need to be a key, the more I think about it. It being a key would prevent derived applications from extending their own interfaces, anyway. Let's just make it a non-protected, boring struct. Then pass by copy in all of the constructors.

NamjaeChoi replied:
> It being a key would prevent derived applications from extending their own interfaces

Can you elaborate this?

loganharbour replied:
Is there a situation in which a derived application would need to also define one of these constructors? If so - they couldn't (due to the friend locking)

NamjaeChoi replied:
Changed to a simple tag.

NamjaeChoi committed:
Addressing comments 8 #30655

lindsayad committed:
Restore passkey for device construction #30655

lindsayad committed:
Use passkey for isKokkosObject #30655

loganharbour reviewed (changes requested):
<img width="318" height="212" alt="image" src="https://github.com/user-attachments/assets/b310d65e-749e-4b03-99c7-5ed9ed332e98" />

I approve; just blocking because I need to make some recipe changes once things are passing

loganharbour reviewed (approved):
Merging despite tests running; running tests are not PR tests and passed previously
